### PR TITLE
RPM: build go.d.plugin native for CentOS-Stream

### PR DESCRIPTION
**Pull request by k0ste**
* enable build_ids for c8s & c9s
* possibility to build netdata in closed environments (without Internet)

Fixed RPM issue
```
TODO: Temporary fix for the build-id error during go.d plugin set up
```

### CODE REVIEW

**Comment by ilyam8** [comment]
Hi, @k0ste. Can you please elaborate on why it is better that way?

**Comment by k0ste** [comment]
> Hi, @k0ste. Can you please elaborate on why it is better that way?

```
The build id is used to identify the debuginfo package that needs to
be installed if gdb is being used to debug the specific library.

It's generated by the build system when the library is compiled. It used
to be that debug packages had to be manually selected, and if more than
one version of a library is available, it was up to the gdb user to get
the right debug packages installed.

The gdb system has been modified, as have the compilers. Instead of
generating debug packages, the debuginfo packages are generated, identified
by the build id. It allows the gdb program to identify the needed debuginfo
packages without the person running gdb having to figure out which packages
are needed.
```
©️[David W. Hodgins](https://alt.os.linux.mageia.narkive.com/u7BL5NiU/what-is-the-directory-usr-lib-build-id)

Second, is not always possible to make builds at the internal CI systems (e.g. gitlab), due lack of internet

File: netdata.spec.in
Comment (Ferroin) [inline netdata.spec.in:223]:
```suggestion
%if 0%{?suse_version}
BuildRequires: go
%else
BuildRequires:  golang
%endif
```

Needed so that these changes work on openSUSE.

**Review by Ferroin — COMMENTED** [review]
> But, need help where I should define `GO_PACKAGE_VERSION`?

Handling for this would need to be added in https://github.com/netdata/helper-images/blob/master/package-builders/fedora-build.sh and https://github.com/netdata/helper-images/blob/master/package-builders/suse-build.sh. I can take care of that though once we’ve confirmed we want this (other changes will be needed in that repo anyway to add the new build dependencies).

**Comment by Ferroin** [comment]
@ilyam8 Functioning debuginfo packages are a major benefit here, and TBH I see no real downsides to taking this approach. Our package builds will take a bit longer, but that’s not a serious problem IMO.

File: netdata.spec.in
Comment (k0ste) [reply netdata.spec.in]: Thanks

**Comment by k0ste** [comment]
> > But, need help where I should define `GO_PACKAGE_VERSION`?
> 
> Handling for this would need to be added in https://github.com/netdata/helper-images/blob/master/package-builders/fedora-build.sh and https://github.com/netdata/helper-images/blob/master/package-builders/suse-build.sh. I can take care of that though once we’ve confirmed we want this (other changes will be needed in that repo anyway to add the new build dependencies).

Reasonable. Accepted

**Comment by k0ste** [comment]
> @ilyam8 Functioning debuginfo packages are a major benefit here, and TBH I see no real downsides to taking this approach. Our package builds will take a bit longer, but that’s not a serious problem IMO.

Ilya, what do you think? 🛳️

**Comment by k0ste** [comment]
Rebased & fixed conflicts

**Comment by k0ste** [comment]
Rebased after merge #14552

**Comment by k0ste** [comment]
@tkatsoulas can you also look this PR please, before another merge conflict 😌

**Comment by ilyam8** [comment]
@k0ste rpm builds fail with the following error

```
error: Failed build dependencies:
	golang is needed by netdata-1.38.0-1.el9.x86_64
```

**Comment by ilyam8** [comment]
> > @ilyam8 Functioning debuginfo packages are a major benefit here, and TBH I see no real downsides to taking this approach. Our package builds will take a bit longer, but that’s not a serious problem IMO.
> 
> Ilya, what do you think? 🛳️

I don't understand how it helps. I'd appreciate a less general description, but at least one use case when/why a user needs that.

**Comment by Ferroin** [comment]
> @k0ste rpm builds fail with the following error
> 
> ```
> error: Failed build dependencies:
> 	golang is needed by netdata-1.38.0-1.el9.x86_64
> ```

This just needs a PR to add it in the appropriate package-builder images in https://github.com/netdata/helper-images. I’ve been holding off on doing so until we determine whether we want the changes here or not.



> > > @ilyam8 Functioning debuginfo packages are a major benefit here, and TBH I see no real downsides to taking this approach. Our package builds will take a bit longer, but that’s not a serious problem IMO.
> > 
> > 
> > Ilya, what do you think? passenger_ship
> 
> I don't understand how it helps. I'd appreciate a less general description, but at least one use case when/why a user needs that.

Without this, the native debugging tooling (not GDB, but the extra stuff provided by Fedora/RHEL) doesn’t properly handle debugging the Go plugin.

I think we may actually need this to have debugging info for the Go plugin at all (not just to link the debuginfo package properly).

**Comment by k0ste** [comment]
> @k0ste rpm builds fail with the following error
> 
> ```
> error: Failed build dependencies:
> 	golang is needed by netdata-1.38.0-1.el9.x86_64
> ```

For `centos:7` build fails, because `golang` package present in EPEL repo. I don't know enabled the `epel` repo on build stage or not. I double check CentOS 7 build with mock, and without EPEL build fails:

```
centos+epel-7-x86_64/root.log
2277:DEBUG util.py:445:   --> golang-1.18.9-1.el7.x86_64
2311:DEBUG util.py:445:  ---> Package golang.x86_64 0:1.18.9-1.el7 will be installed
2312:DEBUG util.py:445:  --> Processing Dependency: golang-bin = 1.18.9-1.el7 for package: golang-1.18.9-1.el7.x86_64
2313:DEBUG util.py:445:  --> Processing Dependency: golang-src = 1.18.9-1.el7 for package: golang-1.18.9-1.el7.x86_64
2336:DEBUG util.py:445:  ---> Package golang-bin.x86_64 0:1.18.9-1.el7 will be installed
2337:DEBUG util.py:445:  ---> Package golang-src.noarch 0:1.18.9-1.el7 will be installed
2412:DEBUG util.py:445:   golang                   x86_64    1.18.9-1.el7               epel       685 k
2430:DEBUG util.py:445:   golang-bin               x86_64    1.18.9-1.el7               epel       105 M
2431:DEBUG util.py:445:   golang-src               noarch    1.18.9-1.el7               epel       9.1 M

centos-7-x86_64/root.log
2208:DEBUG util.py:445:  Error: No Package found for golang
```

Why Alma build was fails - I don't know. The golang package is [present in AppStream](https://almalinux.pkgs.org/8/almalinux-appstream-x86_64/golang-1.18.9-1.module_el8.7.0+3397+4350156d.x86_64.rpm.html) repo

Maybe @Ferroin can help to resolve this

**Comment by k0ste** [comment]
> This just needs a PR to add it in the appropriate package-builder images in https://github.com/netdata/helper-images. I’ve been holding off on doing so until we determine whether we want the changes here or not.

Here the answer @ilyam8, the `builddeps` is hardcoded in docker files. Seems we can't check the build stage, before merge this PR

**Comment by ilyam8** [comment]
> I think we may actually need this to have debugging info for the Go plugin at all (not just to link the debuginfo package properly).

Ok, and why do you need that? Can you please give a specific use case and why there were no requests for it so far?

**Comment by k0ste** [comment]
> why there were no requests for it so far?

Because the RHEL/Fedora official builds comes without go.d.plugin! After last `netdata` releases is almost impossible to use `netdata`, because without `go.d.plugin` we can't even `ping`

[BZ#1818037](https://bugzilla.redhat.com/show_bug.cgi?id=1818037)

**Comment by k0ste** [comment]
If you look to the BZ, it becomes clear that I was trying to persuade the CentOS package maintainer to add `go.d.plugin` support, because it was the only option not to do my own build. Then we made our Gitlab CI build. But in this time @Ferroin introduced cool `netdata` RPM repositories, so it would be nice to make a RHEL compatible build here (I think)

**Comment by ilyam8** [comment]
> If you look to the BZ, it becomes clear that I was trying to persuade the CentOS package maintainer to add go.d.plugin support, because it was the only option not to do my own build.

Thanks, @k0ste. Finally, I get a good explanation. I don't follow RHEL/Fedora official builds and wasn't aware they don't include go.d.plugin because of that reason.

**Comment by k0ste** [comment]
Rebased again

File: netdata.spec.in
Comment (Ferroin) [inline netdata.spec.in:70]: We officially do not support RHEL 6, so we don’t need special handling of it here.

Also, unless I’m mistaken, this appears to actually disable the Go plugin everywhere, which would explain why CI appears to be passing even though we never actually got the needed changes added to the package builder images (my bad on that one, I’ll jump on that later today).

Comment (k0ste) [reply netdata.spec.in]: RHEL 6 condition removed

**Comment by ilyam8** [comment]
@k0ste please prefer commits over force-push because it is not possible to track the changes. We do squash before merging.

**Comment by Ferroin** [comment]
Overall LGTM. 

I’m currently working on our end on getting the package builder changes in https://github.com/netdata/helper-images/pull/203 merged (should be some time later today) at which point I’ll manually re-trigger the CI to confirm that the package builds work correctly, but assuming there are no issues there we should be able to merge this some time this week so that it ends up in the v1.39.0 stable release.

**Comment by k0ste** [comment]
> Overall LGTM.
> 
> I’m currently working on our end on getting the package builder changes in [netdata/helper-images#203](https://github.com/netdata/helper-images/pull/203) merged (should be some time later today) at which point I’ll manually re-trigger the CI to confirm that the package builds work correctly, but assuming there are no issues there we should be able to merge this some time this week so that it ends up in the v1.39.0 stable release.

Sounds good, thanks!

File: netdata.spec.in
Comment (Ferroin) [inline netdata.spec.in:74]: Not sure how I missed this previously.

```suggestion
# go.d.plugin
%bcond_without goplugin
```

RPM will automatically provide the required logic for a CLI option for `rpmbuild` to disable the plugin, and will default to having it enabled.

**Comment by k0ste** [comment]
Rebased again, after global changes in SPEC

**Comment by k0ste** [comment]
Fixed another conflict after merges in SPEC 🌚

**Comment by Ferroin** [comment]
The required changes were merged in the helper-images repo, and the package builder images have all been published, so I’ve retriggered the package build checks on this PR.

**Review by Ferroin — DISMISSED** [review]

**Comment by k0ste** [comment]
In general, until it was merged, the plugin version was set required golang to 1.20, now building on non-CentOS Stream RPM distributions is not possible (lack of golang 1.20). We added CentOS Stream images https://github.com/netdata/helper-images/pull/219, now we should merge CI changes https://github.com/netdata/netdata/pull/15742 to check this PR

**Comment by k0ste** [comment]
CI changes merged, for c8s build passed with native go build

**Review by Ferroin — DISMISSED** [review]

**Comment by k0ste** [comment]
Please review before the next conflict 🙏

**Review by Ferroin — APPROVED** [review]

File: netdata.spec.in
Comment (tkatsoulas) [inline netdata.spec.in:913]: This will work at the moment. But I think we need to introduce long term the netdata_user and netdata_group vars

**Review by tkatsoulas — APPROVED** [review]

File: netdata.spec.in
Comment (k0ste) [reply netdata.spec.in:913]: I suggest to totally disable this "old school" `groupadd` and strongly use [systemd-sysusers](https://www.freedesktop.org/software/systemd/man/sysusers.d.html#) for all user/groups tasks.

Why: when `netdata` installed, for example before ceph or docker or nginx, etc package
See: https://github.com/netdata/netdata/blob/master/netdata.spec.in#L461-L466
Results: netdata modules for this services will not work, because if group is not present, netdata user will not be added to this group

We a cover this locally with rules like this:

```shell
[root@ceph-mon0 /]# cat /etc/sysusers.d/99-netdata.conf
m netdata ceph - - -
```

Benefit: package maintainers can use tmpfiles,sysusers shipped with tarball, [rather than doing them manually](https://gitlab.archlinux.org/archlinux/packaging/packages/netdata/-/blob/main/netdata.sysusers?ref_type=heads). When package get more quality and less errors - netdata, as product, performed much better for end-user

Comment (Ferroin) [reply netdata.spec.in:913]: > I suggest to totally disable this "old school" groupadd and strongly use [systemd-sysusers](https://www.freedesktop.org/software/systemd/man/sysusers.d.html#) for all user/groups tasks.

I had actually started to look into this about two weeks ago before I got sidetracked (again) by other projects. It’s something that we do, indeed, almost certainly want to do in cases when we can reliably determine that we’re on a systemd-based system (and the native RPM packages definitely fit that description).

**Pull request merged**
